iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+    PriorityQueueEntry<RefCountCacheHashEntry *> *expiry_entry = expiryQueueEntry.alloc();
+    new ((void *)expiry_entry) PriorityQueueEntry<RefCountCacheHashEntry *>(val);
+    expiry_queue.push(expiry_entry);
+    val->expiry_entry = expiry_entry;
+  }
+
+  // add the item to the map
+  this->item_map.insert(val);
+  this->size += val->meta.size;
+  this->items++;
+  this->metric_inc(refcountcache_current_size_stat, (int64_t)val->meta.size);
+  this->metric_inc(refcountcache_current_items_stat, 1);
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::erase(uint64_t key, int expiry_time)
+{
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    if (expiry_time >= 0 && l.m_value->meta.expiry_time != expiry_time) {
+      return;
+    }
+    // found
+    this->item_map.remove(l);
+
+    // decrement usage counters
+    this->size -= l.m_value->meta.size;
+    this->items--;
+
+    this->metric_inc(refcountcache_current_size_stat, -((int64_t)l.m_value->meta.size));
+    this->metric_inc(refcountcache_current_items_stat, -1);
+
+    // remove from expiry queue
+    if (l.m_value->expiry_entry != NULL) {
+      Debug("refcountcache", "partition %d deleting item from expiry_queue idx=%d\n", this->part_num,
+            l.m_value->expiry_entry->index);
+      this->expiry_queue.erase(l.m_value->expiry_entry);
+      expiryQueueEntry.free(l.m_value->expiry_entry);
+      l.m_value->expiry_entry = NULL; // To avoid the destruction of `l` calling the destructor again-- and causing issues
+    }
+    // Since the Value is actually RefCountObj-- when this gets deleted normally it calls the wrong
+    // `free` method, this forces the delete/decr to happen with the right type
+    Ptr<C> *tmp = (Ptr<C> *)&l.m_value->item;
+    tmp->clear();
+    refCountCacheHashingValueAllocator.free(l.m_value);
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::clear()
+{
+  this->metric_inc(refcountcache_current_size_stat, -this->size);
+  this->metric_inc(refcountcache_current_items_stat, -this->items);
+  // Clear the in memory hashmap
+  // TODO: delete all items (not sure if clear calls delete on all items)
+  this->item_map.clear();
+  // clear the queue
+  PriorityQueueEntry<RefCountCacheHashEntry *> *tmp_entry;
+  while (!this->expiry_queue.empty()) {
+    tmp_entry = this->expiry_queue.top();
+    this->expiry_queue.pop();
+    expiryQueueEntry.free(tmp_entry);
+  }
+
+  this->items = 0;
+  this->size  = 0;
+}
+
+// Are we full?
+template <class C>
+bool
+RefCountCachePartition<C>::is_full()
+{
+  Debug("refcountcache", "partition %d is full? items %d/%d size %ld/%ld\n\n", this->part_num, this->items, this->max_items,
+        this->size, this->max_size);
+  return (this->max_items > 0 && this->items >= this->max_items) || (this->max_size > 0 && this->size >= this->max_size);
+}
+
+// Attempt to make space for item of `size`
+template <class C>
+bool
+RefCountCachePartition<C>::reserve(unsigned int size)
```

Comment:
Ah sorry about that. It is really hard to keep track over everything :-/

This code doesn't do what I would expect ``reserve`` to do. I'm not convinced that you need to know ``size`` here at all; just evict everything that has expired.